packages/@ember/array/type-tests/index.ts

```diff
@@ -0,0 +1,214 @@
+}, target);
+
+expectTypeOf(arr.filterBy('bar')).toEqualTypeOf<NatifiedArray<Foo>>();
+expectTypeOf(arr.filterBy('missing')).toEqualTypeOf<NatifiedArray<Foo>>();
+
+expectTypeOf(arr.rejectBy('bar')).toEqualTypeOf<NatifiedArray<Foo>>();
+expectTypeOf(arr.rejectBy('missing')).toEqualTypeOf<NatifiedArray<Foo>>();
+
+expectTypeOf(arr.findBy('bar')).toEqualTypeOf<Foo | undefined>();
+arr.findBy('bar', 1);
+// @ts-expect-error value has incorrect type
+arr.findBy('bar', 'invalid');
+arr.findBy('missing', 'whatever');
+
+let isEvery = arr.every((item, index, arr) => {
+  expectTypeOf(item).toEqualTypeOf<Foo>();
+  expectTypeOf(index).toEqualTypeOf<number>();
+  expectTypeOf(arr).toEqualTypeOf(arr);
+  return true;
+});
+expectTypeOf(isEvery).toEqualTypeOf<boolean>();
+arr.every(function (_item) {
+  // @ts-expect-error the EmberArray version isn't recognized
+  expectTypeOf(this).toEqualTypeOf(target);
+  return true;
+}, target);
+
+expectTypeOf(arr.isEvery('bar')).toEqualTypeOf<boolean>();
+arr.isEvery('bar', 1);
+// @ts-expect-error value has incorrect type
+arr.isEvery('bar', 'invalid');
+arr.isEvery('missing', 'whatever');
+
+let isAny = arr.any((item, index, arr) => {
+  expectTypeOf(item).toEqualTypeOf<Foo>();
+  expectTypeOf(index).toEqualTypeOf<number>();
+  expectTypeOf(arr).toEqualTypeOf(arr);
+  return true;
+});
+expectTypeOf(isAny).toEqualTypeOf<boolean>();
+arr.any(function (_item) {
+  expectTypeOf(this).toEqualTypeOf(target);
+  return true;
+}, target);
+
+expectTypeOf(arr.isAny('bar')).toEqualTypeOf<boolean>();
+arr.isAny('bar', 1);
+// @ts-expect-error value has incorrect type
+arr.isAny('bar', 'invalid');
+arr.isAny('missing', 'whatever');
+
+let reduced = arr.reduce((summation, item, index, arr) => {
+  expectTypeOf(summation).toEqualTypeOf<number>();
+  expectTypeOf(item).toEqualTypeOf<Foo>();
+  expectTypeOf(index).toEqualTypeOf<number>();
+  expectTypeOf(arr).toEqualTypeOf(arr);
+  return 1;
+}, 1);
+expectTypeOf(reduced).toEqualTypeOf<number>();
+// Native behavior takes precedence here
+expectTypeOf(arr.reduce((summation, _item) => summation)).toEqualTypeOf<Foo>();
+
+expectTypeOf(arr.invoke('hi')).toEqualTypeOf<NatifiedArray<string>>();
+expectTypeOf(arr.invoke('withArgs', 1, 'two')).toEqualTypeOf<NatifiedArray<string>>();
+// @ts-expect-error Doesn't allow calling with invalid args
+arr.invoke('withArgs', 'invalid');
+expectTypeOf(arr.invoke('missing')).toEqualTypeOf<NatifiedArray<unknown>>();
+// Currently, args passed to unrecognized methods are ignored
+arr.invoke('missing', 'invalid');
+
+expectTypeOf(arr.toArray()).toEqualTypeOf<Foo[]>();
+
+expectTypeOf(arr.compact()).toEqualTypeOf<NatifiedArray<Foo>>();
+expectTypeOf(A([foo, null]).compact()).toEqualTypeOf<NatifiedArray<Foo>>();
+
+expectTypeOf(arr.includes(foo)).toEqualTypeOf<boolean>();
+// @ts-expect-error Invalid type
+arr.includes(1);
+
+// For some reason this doesn't return a NatifiedArray
+expectTypeOf(arr.sortBy('bar')).toEqualTypeOf<Foo[]>();
```

Comment:
As with the others, does it work with `toMatchTypeOf`?

```suggestion
expectTypeOf(arr.sortBy('bar')).toMatchTypeOf<NativeArray<Foo>>();
```

Reply:
Maybe. I'll review this as well.